### PR TITLE
fix(#3015): dynamic width no longer truncates on right_align icons

### PR DESCRIPTION
**Pull request by simonmandlik**
Fixes https://github.com/nvim-tree/nvim-tree.lua/issues/3015

### CODE REVIEW

**Comment by alex-courtis** [comment]
Thanks for the fix, however I'm having trouble replicating and validating it.

Please provide a [clean room configuration](https://github.com/nvim-tree/nvim-tree.lua/wiki/Development#clean-room) and instructions on how I can demonstrate the fix with it.

**Comment by simonmandlik** [comment]
Instructions:

1. run `nvim -nu minimal.lua`
2. Edit the longest file in any git repo. Here, I use `nvim-tree` directly so I change something in `.release-please-manifest.json` so that it get a right-aligned icon
3. This is the bug

image

Even though the max width is set to 100%, the whole filename is not shown due to the icon covering a part of it.

With this PR, it looks like this

image

`minimal.lua`:
```lua
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvt-min/site]])
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup({
    {
      "wbthomason/packer.nvim",
      -- "~/Downloads/nvim-tree.lua",
      "nvim-tree/nvim-tree.lua",
      "nvim-tree/nvim-web-devicons",
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  })
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing nvim-tree and dependencies.")
  vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]])
vim.opt.termguicolors = true
vim.opt.cursorline = true

_G.setup = function()
  require("nvim-tree").setup({
    view = {
      width = {
        min = -1,
        max = "100%",
        padding = 0
      }
    },
    renderer = {
      icons = {
        git_placement = "right_align"
      }
    }
  })
end
```

**Comment by simonmandlik** [comment]
Addressed the CI issues

**Comment by alex-courtis** [comment]
This change does address the issue for width whilst focused. Unfortunately the width when changing is a regression.

View is quite fragile. It is targeted for a refactor during https://github.com/orgs/nvim-tree/projects/1 however that will not address this issue.

A full solution is desirable, let's see what we can come up with.

File: lua/nvim-tree/view.lua [hunk 1]
```diff
@@ -321,8 +321,19 @@ local function grow()
     max_width = get_width(M.View.max_width) - padding
   end
 
-  for _, l in pairs(lines) do
+  local ns_id = vim.api.nvim_get_namespaces()["NvimTreeExtmarks"]
+  for line_nr, l in pairs(lines) do
     local count = vim.fn.strchars(l)
+    -- also add space for right-aligned icons
+    local extmarks = vim.api.nvim_buf_get_extmarks(0, ns_id, { line_nr, 0 }, { line_nr, -1 }, { details = true })
```
Comment (alex-courtis) [inline lua/nvim-tree/view.lua:328]: This looks like it might be the issue: we're getting extmarks for the current buffer rather than the tree buffer.

Comment (simonmandlik) [reply lua/nvim-tree/view.lua]: Thank you, fixed!

**Review by alex-courtis — APPROVED** [review]
Works nicely for a variety of width cases.
Many thanks for resolving this one!

**Pull request merged**